views/build/grunt/portableelement.js

```diff
@@ -0,0 +1,219 @@
+            minHookFile = pciRuntimeData.hook;
+        } else if (Array.isArray(pciRuntimeData.libraries) && pciRuntimeData.libraries.length > 0) {
+            //by convention the first module is the min file
+            minHookFile = pciRuntimeData.libraries[0];
+        }
+        if(minHookFile){
+            return minHookFile.replace(/^\.\//, '');
+        }
+    }
+
+    /**
+     * Print report when all promises are resolved/rejected
+     * @param {Array|Function} report
+     */
+    function printReport(report) {
+        if(Array.isArray(report)){
+            report.forEach(function (r) {
+                printReport(r);
+            });
+        }else if(typeof report === 'function'){
+            report.call();
+        }
+    }
+
+    /**
+     * Get the portable element model from its manifest file
+     * @param file
+     * @returns {*}
+     */
+    function getPortableModelFromFile(file){
+        var model;
+        portableModels.forEach(function(portableModel){
+            if(file.match(new RegExp('\/' + portableModel.file + '$'))){
+                model = {};
+                model.type = portableModel.type;
+                model.manifest = grunt.file.readJSON(file);
+                model.basePath = file.replace('/' + portableModel.file, '');
+                model.id = model.manifest.typeIdentifier;
+                model.runtimeHook = getHookFileName(model.manifest.runtime, model.id);
+                model.minRuntimeFile = getMinHookFile(model.manifest.runtime);
+            }
+        });
+        return model;
+    }
+
+    grunt.registerTask('portableelement', 'Compile Portable Elements', function () {
+
+        var done = this.async();//async mode because requirejs optimization is an async process
+        var extension = grunt.option('extension') || grunt.option('e');
+        var selectedId = grunt.option('identifier') || grunt.option('i');
+        var manifests, compileTasks;
+        var self = this;
+
+        if(!extension){
+            grunt.log.error('Missing the extension in param, e.g. "grunt portableelement -e=qtiItemPci"');
+            return done();
+        }
+
+        grunt.log.writeln('Started optimizing portable elements in extension "' + extension + '"');
+
+        if(selectedId){
+            grunt.log.writeln('Only searching portable element "' + selectedId + '"');
+        }
+
+        manifests = portableModels.reduce(function(acc, model) {
+            return grunt.file.expand(root + '/' + extension + model.searchPattern).concat(acc);
+        }, []);
+
+        compileTasks = manifests.map(function(file){
+            return new Promise(function (resolve, reject) {
+                var report = [];
+                var config;
+                var model = getPortableModelFromFile(file);
+
+                if(!model){
+                    //not the targeted one
+                    return resolve([grunt.log.error.bind(null, 'invalid portable manifest file ' + file)]);
```

Comment:
Why do you use `.bind` on all log methods ? Are the methods expecting to run in the lexical scope where `this = grunt` ?

Reply:
It is for the messages to stay grouped and ordered while each async optimization action run.
With a direct call, the messages order are messed up so we can get one message from one compilation task under another one...